File: 82. Remove Duplicates from Sorted List II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        dummyhead = ListNode(head.val-1)
        dummyhead.next = head
        prev = dummyhead
        while head and head.next:
            if  head.val == head.next.val:
                while head.next and head.val == head.next.val:
                    head.next = head.next.next
                prev.next = head.next
                head = prev.next
            else:
                prev = head
                head = head.next
        return dummyhead.next

File: 82. Remove Duplicates from Sorted List II.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode node = head, prev = dummy;
        while (node != null) {
            while (node.next!=null && node.val == node.next.val) {
                node = node.next;
            }
            if (prev.next != node) {
                prev.next = node.next;
                node = node.next;
            } else {
                prev = node;
                node = node.next;
            }
        }
        return dummy.next;
    }
}